BSFER-73 Add event_lead_image module.

tschanzt opened the pull request:
Adds the event_lead_image module to add a lead image in multiple sizes.
Add Form:
![screenshot from 2018-02-19 12-18-32](https://user-images.githubusercontent.com/358342/36375801-0a9155c6-1571-11e8-9f79-8a1b3abe5236.png)

Commit:
Add event_lead_image module.

simahawk commented on event_lead_image/README.rst:
```diff
@@ -0,0 +1,50 @@
+   :target: https://www.gnu.org/licenses/agpl
+   :alt: License: AGPL-3
+
+============================
```
RST wants wrapping lines to match text length

simahawk commented on event_lead_image/README.rst:
```diff
@@ -0,0 +1,50 @@
+Event Lead Image
+============================
+
+This module extends the functionality of event.
```
drop 1st line

simahawk commented on event_lead_image/README.rst:
```diff
@@ -0,0 +1,50 @@
+
+This module extends the functionality of event.
+This Module adds the `lead_image` field to the event
+and calculates it fixed size versions from it. It is available in the sizes big, medium, small.
```
"and provides scaled size versions (big, medium, small)."

simahawk commented on event_lead_image/models/event.py:
```diff
@@ -0,0 +1,42 @@
+            if not self._context.get('bin_size'):
+                item.update(item._get_resized_images())
+            else:
+                self.lead_image_big = self.lead_image
```
then you set the same big image on each? explain the `bin_size` story pls ;)

simahawk commented on event_lead_image/tests/test_lead_image.py:
```diff
@@ -0,0 +1,41 @@
+    def setUpClass(cls):
+        super().setUpClass()
+        path = os.path.dirname(__file__) + '/assets/testimage.png'
+        file_ = open(path, 'rb')
```
use a ctx manager to close the file when not needed anymore
```python
with open(path, 'rb') as img_file:
    image = b64encode(img_file_.read())
```

simahawk commented on event_lead_image/tests/test_lead_image.py:
```diff
@@ -0,0 +1,41 @@
+    def test_access_binsize(self):
+        """Test that we don't fail if we're in bin_size mode."""
+        event = self.env['event.event'].with_context(
+            bin_size=True).browse(self.testevent.id)
```
comment: `bin_size` make the fields return images' sizes instead of image content

simahawk commented on event_lead_image/models/event.py:
```diff
@@ -0,0 +1,42 @@
+    @api.depends("lead_image")
+    def _compute_images(self):
+        for item in self:
+            if not self._context.get('bin_size'):
```
never use private stuff unless you are forced to: `self.env.context`

simahawk commented on event_lead_image/views/event_views.xml:
```diff
@@ -0,0 +1,20 @@
+            <field name="inherit_id" ref="event.view_event_form" />
+            <field name="arch" type="xml">
+
+                <xpath expr="//div[@class='oe_title']" position="attributes">
```
never select classes like this, you can have more than one and the xpath would fail. If you check your instance's logs I bet you'll find a warning for it saying that you should use `hasclass('oe_title')` :wink:

simahawk replied:
@tschanzt what about this?

tschanzt replied:
@simahawk ah now i realize what you meant.Somehow i managed to mangle your sentence into you complaining about there possibly being more than one result. I think i used this way quite a lot till now :scream:

simahawk commented on event_lead_image/views/event_views.xml:
```diff
@@ -0,0 +1,20 @@
+                    <attribute name="style">float: left; width: 100%</attribute>
+                </xpath>
+                <xpath expr="//div[@class='oe_title']" position="after">
+                    <div class="oe_leadimage" style="float:left;">
```
`class="o_leadimage pull-right"` and drop `style`

simahawk commented on event_lead_image/views/event_views.xml:
```diff
@@ -0,0 +1,20 @@
+                    <div class="oe_leadimage" style="float:left;">
+                        <label for="lead_image" string="Lead Image" style="display: block;" />
+                        <field name="lead_image" class="oe_avatar" widget="image" />
+                        <field name="lead_image_small" />
```
why 2 images?

simahawk commented on event_lead_image/views/event_views.xml:
```diff
@@ -0,0 +1,20 @@
+                </xpath>
+                <xpath expr="//div[@class='oe_title']" position="after">
+                    <div class="oe_leadimage" style="float:left;">
+                        <label for="lead_image" string="Lead Image" style="display: block;" />
```
label not needed, if you want it float it as well

tschanzt commented:
@simahawk i added a short comment about the bin_size stuff but here is a bit a more extensive version. As far as i can tell is only called when the image is needed and it doesn't have a value yet. Now if it is called with the bin_size flag i should return the size of the image which doesn't exist yet. so i use the one of the original image. 

I also addressed the rest of your comments.

simahawk commented on event_lead_image/tests/test_lead_image.py:
```diff
@@ -0,0 +1,43 @@
+        # actual image data.
+        event = self.env['event.event'].with_context(
+            bin_size=True).browse(self.testevent.id)
+        self.assertAlmostEqual(self.convert_to_float(event.lead_image),
```
how comes that the resized smaller images have the same size? :)

tschanzt commented:
@simahawk updated again.

ghost commented:
is it really related to [BSFER-69](https://jira.camptocamp.com/browse/BSFER-69), as the title says?
looks more like [BSFER-73](https://jira.camptocamp.com/browse/BSFER-73) :)

arkostyuk commented on event_lead_image/tests/test_lead_image.py:
```diff
@@ -0,0 +1,47 @@
+                               114.59)
+
+    def convert_to_float(self, size_string):
+        return float(size_string.split(" ")[0])
```
just FYI, `split()` does the same as `split(' ')`

arkostyuk commented on event_lead_image/views/event_views.xml:
```diff
@@ -0,0 +1,18 @@
+<odoo>
```
misses an encoding comment

tschanzt commented:
@arkostyuk @simahawk updated.

simahawk commented on event_lead_image/tests/test_lead_image.py:
```diff
@@ -0,0 +1,47 @@
+
+        # bin_size makes the attachment return its size instead of the
+        # actual image data.
+        event = self.env['event.event'].with_context(
```
FYI this can be done like: `event = self.testevent.with_context(bin_size=True)`
